gohcl/decode_test.go

```diff
@@ -512,13 +511,220 @@ func TestDecodeBody(t *testing.T) {
 				"name":   "Ermintrude",
 				"living": true,
 			},
-			map[string]cty.Value(nil),
+			makeInstantiateType(map[string]cty.Value(nil)),
 			deepEquals(map[string]cty.Value{
 				"name":   cty.StringVal("Ermintrude"),
 				"living": cty.True,
 			}),
 			0,
 		},
+		{
+			// Retain "nested" block while decoding
+			map[string]interface{}{
+				"foo": map[string]interface{}{
+					"plain": "foo",
+				},
+			},
+			func() interface{} {
+				return reflect.ValueOf(&struct {
```

Comment:
I think the `reflect.ValueOf` shouldn't be needed here, because the value you want to return is fixed. The existing cases were using `reflect` to construct a zero value for a given type, but since we don't want a zero value here I think it should work to just write out the literal directly:

```go
return &struct {
    Foo struct { ... } `hcl:"foo,block"`
}{
    Foo: struct {
        ...
    }
}
```

With that said, these structs are complex enough that I think it's worth making a named type like the `type withNameExpression struct` above to make this easier to read.

e.g. if we define struct types like this:

```go
type withTwoAttributes struct {
    A string `hcl:"a,optional"`
    B string `hcl:"b,optional"`
}

type withNestedBlock struct {
    Plain  string             `hcl:"plain,optional"`
    Nested *withTwoAttributes `hcl:"nested,block"`
}
```

...then this function would have a lot less "type noise" in it, making it easier to read what the values are:

```go
func interface{} {
    &withNestedBlock{
        Plain: "bar",
        Nested: &withTwoAttributes{
            B: "bar",
        },
    }
}
```

It seems like there's only two or three distinct types across all of these new test cases, so hopefully it shouldn't take too many struct declarations to cover them all, since we can share the same type across multiple if they are close enough for what the test case is trying to verify.

Reply:
Yeah, I got it way too complicated. Thanks for suggesting the simplification. Will push updated version soon.